src/backend/distributed/planner/deparse_shard_query.c

```diff
@@ -100,7 +100,16 @@ RebuildQueryStrings(Query *originalQuery, List *taskList)
 
 		ereport(DEBUG4, (errmsg("query before rebuilding: %s", task->queryString)));
 
-		UpdateTaskQueryString(query, relationId, valuesRTE, task);
+		/* we need to use copy of the query if it is a multi shard update/delete query */
+		if (UpdateOrDeleteQuery(query) && list_length(taskList))
```

Comment:
The problem this seems to solve is that we need to copy the query tree when there are multiple tasks, such that we start from the original every time.

What if we do this at the start?

```
/* when there are multiple tasks, start from a fresh copy of the query tree */
if (list_length(taskList) > 0)
{
    query = copyObject(originalQuery);
}
```

then in the `INSERT...SELECT` logic we can remove the `query = copyObject(originalQuery);` line.